content/blog/2021-07-06-hyperparam-tuning.md

```diff
@@ -0,0 +1,241 @@
+manually updating a config file with new hyperparameter values for each
+experiment run. The command syntax for creating queues looks like this:
+
+`dvc exp run --queue -set-param train.min_split=8`
+
+In the example queue above, we're updating the `train.min_split` value that's
+inside of the `params.yaml` file. This file holds all of the hyperparameter
+values and is where DVC looks to determine if any values have changed. With the
+command above, we're automatically updating that value in the `params.yaml`
+using a queued experiment.
+
+Now we can make the script. You can add a new file to the `src` directory called
+`tuning.py`. Inside of the file, add the following code.
+
+```python
+import subprocess
+import random
+
+# Automated grid search experiments
+n_est_values = [250, 300, 350, 400, 450, 500]
+min_split_values = [100, 110, 120, 130, 140, 150]
+
+for val in n_est_values:
+    subprocess.run(["dvc", "exp", "run", "--queue", "--set-param", f"train.n_est={val}"])
+
+for val in min_split_values:
+    subprocess.run(["dvc", "exp", "run", "--queue", "--set-param", f"train.min_split={val}"])
+```
+
+This is a simple grid search. We have two hyperparameters we want to tune:
+`n_est` and `min_split`. So we have arrays with a few values in them to mimic
+the exhaustive search a grid search can handle. Then we loop through the values
+and create queued experiments for them using `subprocess`.
+
+You can run this script now and generate your queue with this command.
+
+`python src/tuning.py`
+
+You'll see some outputs in the terminal telling you that your experiments have
+been queued. Then you can run them all with the following command.
+
+`dvc exp run --run-all`
+
+This will run every experiment that has been queued. Once all of those have run,
+take a look at your metrics for each experiment.
+
+`dvc exp show`
+
+Your table should look similar to this when you run `dvc exp show`.
+
+```bash
+table
+```
+
+Now you can see how your precision changed with each hyperparameter value
+update. This is a quick implementation of grid search in DVC. You could read the
+hyperparameter values from a different file or data source or make this tuning
+script as fancy as you like. The main thing you need is the
+`dvc exp run --queue -set-param <param>` command to execute when you add new
+values.
+
+### Random Search
+
+Another commonly used method for tuning hyperparameters is random search. This
+takes random values for hyperparameters and builds the model with them. It's not
+an exhaustive search like grid search and it typically outperforms grid search,
+especially when a small subset of the hyperparameters produce the biggest
+difference in your model.
+
+We're going to add an example of random search to the `tuning.py` file we
+created for grid search. This will add queued experiments with the randomly
+selected hyperparameter values. Add the following code to `tuning.py` below the
+grid search implementation.
+
+```python
```

Comment:
The comments from the code block above apply here, too.

Also, I would recommend randomly choosing each hyperparameter on each iteration instead of choosing a random set of values for each hyperparameter beforehand. For example:

```
for _ in num_exps:
    n_est = random.randint(250, 500)
    min_split = random.choice([8,16,32,64,128,256]) # You could try to think of more clever ways to expand on this, although we probably want to avoid numpy functions
```